### PR TITLE
Timestamp normalize

### PR DESCRIPTION
**Pull request by nick-Ag**
Adds normalization of both axis to the app. Users can place all plots on top of each other, or normalize them in [-1, 1] to compare possibly related datasets. Since calculations for this are simple (adding/subtracting the same value from every datapoint), they are done in the frontend rather than making another backend call. However, the backend does pre-compute the min and max value of all traces since this will be needed for y-axis normalization and would be too expensive to perform in the frontend.

### CODE REVIEW

File: frontend/src/app/dataset/dataset.component.ts [hunk 1]
```diff
@@ -275,10 +286,87 @@ export class DatasetComponent {
...
+  normalizeYHelper(i: string, normalize: boolean, new_data: Array<number>) {
+    this.sample.data[i]['arr'].forEach((value, index) => {
+      if (value === 0) {
+        new_data[index] = 0;
+      } else {
+        const [min, max] = this.sample.data[i]['minmax'];
+        if (normalize) {
+          new_data[index] = (max - min) * ((value + 1) / 2) + min;
+        } else {
+          new_data[index] = 2 * ((value - min) / (max - min)) - 1;
```
Comment (rglasser) [inline frontend/src/app/dataset/dataset.component.ts:366]: What if `max == min` ?

File: backend/parser.py [hunk 2]
```diff
@@ -271,23 +271,34 @@ def jsonify(self, samples: list):
         ret_dict = {}
 
         for i, sample in enumerate(samples):
-            # sample.data[key][0] is a placeholder for the unique id
+            # sample.data[key]['id'] is a placeholder for the unique id
```
Comment (rglasser) [inline backend/parser.py:274]: Can you document 'minmax' and 'arr' as well?

File: frontend/src/app/dataset/dataset.component.ts [hunk 3]
```diff
@@ -342,7 +342,18 @@ export class DatasetComponent {
     this.normalizationY = !this.normalizationY;
     for (const i in this.sample.data) {
       const new_data = new Array<number>(this.sample.data[0]['arr'].length);
-      this.normalizeYHelper(i, !this.normalizationY, new_data);
+      const [min, max] = this.sample.data[i]['minmax'];
+      // Edge case where a straight line trace would cause a divide by 0.
+      // If a trace with all 0s occurs, it is not normalized in normalizeYHelper.
+      if (min !== 0 && min === max) {
```
Comment (rglasser) [inline frontend/src/app/dataset/dataset.component.ts:348]: I would put this part in `normalizeYHelper` as well. Currently this fragments the part of the code.

**Review by rglasser — APPROVED** [review]

File: frontend/src/app/side-menu/side-menu.component.ts [hunk 4]
```diff
@@ -69,6 +72,7 @@ export class SideMenuComponent {
           }
           case 'upload': {
             // Plot new datasets added by the upload button.
+            console.log('smenu dashboard', this.dashboard);
```
Comment (agarwal-v) [inline frontend/src/app/side-menu/side-menu.component.ts:75]: Is this helpful?

**Review by agarwal-v — APPROVED** [review]
1 minor comment.

**Pull request merged**